### DISCUSSION TITLE
[WEEK4 補充資料] 關於非同步的問題  race condition

### DISCUSSION THREAD

**Discussion by raye0621** [Lidemy/forum#96] [其他問題] [answered]
week 4 的補充資料中有提到，要避免 race condition 的狀況產生，以下範例可以這樣寫：
```javascript
const request = require('request');

// 新增留言
request.post({
  url: 'https://example.com/api/messages',
  form: { title: '新留言' }
}, (err, res) => {
  console.log('新增成功！');

  // 確保新增成功，才去抓取所有留言
  request('https://example.com/api/messages', (err, res, body) => {
    console.log('所有留言：', body)
  })
})
```

那...為什麼這樣就可以避免呢？
把抓取留言包含在新增留言的 Request 中，這樣 Request 過去的話他是會先回來再 Request 過去嗎

有點像繞口令...
想要請問這個 request 的函式是這樣運作的嗎?
![Q1](https://i.imgur.com/us2LvV8.jpg)

1. 所以等於是在 request 裡面發送 request，一定會等上一個請求回來了，才會發送下一個請求這樣子嗎？
2. 這就是一種寫法而已嗎，所以重點就是把要 「後執行」 的 Request 包含在要 「先執行」 的裡面就可以了嗎？

感謝解答，如果是我前面漏看了甚麼部分再麻煩提醒一下QQ 謝謝。

**Comment by allenliaooooooo** [comment, marked as answer]
因為新留言那個 request 的函式是拿到 response 才會執行

```

request.post({
  url: 'https://example.com/api/messages',
  form: { title: '新留言' }
}, (err, res) => {
  // ----------------------------------
  // 會執行到這個函式裡面一定是拿到 response 了
  // ----------------------------------
  console.log('新增成功！');

  // 確保新增成功，才去抓取所有留言
  request('https://example.com/api/messages', (err, res, body) => {
    console.log('所有留言：', body)
  })
})

```

-----
### 補充說明
`request.post()` 是一個 function，讓我們可以送出 HTTP request
第一個參數可以傳入 URL, header 之類的資訊
第二個參數傳入一個 function，當我們的 request 得到回應（response）時才會執行這個 function
所以當我們把一個 request 放在這個被傳入的 function 裡面，就可以確保這個 request 會在第一個 request 得到回應（新增留言成功）時，才送出

**Reply by raye0621** [reply]
哦哦哦！我懂了，原本以為是整個 function 裡面的資料都是要傳過去的，現在才知道不是這樣，
原來是我對 `request.post()` 的構造不夠熟悉，現在瞭解了！非常感謝你 :DDD
